src/main/java/com/amazon/opendistroforelasticsearch/sql/expression/core/ExpressionFactory.java

```diff
@@ -0,0 +1,143 @@
+            public String toString() {
+                return String.format("add(%s,%s)", left, right);
+            }
+        };
+    }
+
+    public static Expression sub(Expression left, Expression right) {
+        return new Expression() {
+            @Override
+            public ExprValue valueOf(BindingTuple tuple) {
+                return arithmetic(ArithmeticOperation.ADD, left.valueOf(tuple), right.valueOf(tuple));
+            }
+
+            @Override
+            public String toString() {
+                return String.format("sub(%s,%s)", left, right);
+            }
+        };
+    }
+
+    public static Expression log(Expression expr) {
+        return new Expression() {
+            @Override
+            public ExprValue valueOf(BindingTuple tuple) {
+                final ExprValue exprValue = expr.valueOf(tuple);
+                switch (exprValue.kind()) {
+                    case INTEGER_VALUE:
+                        return doubleValue(Math.log(exprValue.numberValue().intValue()));
+                    case DOUBLE_VALUE:
+                        return doubleValue(Math.log(exprValue.numberValue().doubleValue()));
+                    default:
+                        throw new RuntimeException("unsupported log operand: " + exprValue.kind());
+                }
+            }
+
+            @Override
+            public String toString() {
+                return String.format("log(%s)", expr);
+            }
+        };
+    }
+
+    public static Expression cast(Expression expr) {
+        return new Expression() {
+            @Override
+            public ExprValue valueOf(BindingTuple tuple) {
+                return expr.valueOf(tuple);
+            }
+
+            @Override
+            public String toString() {
+                return String.format("cast(%s)", expr);
+            }
+        };
+    }
+
+    private static ExprValue arithmetic(ArithmeticOperation op, ExprValue v1, ExprValue v2) {
+        if (v1.kind() != v2.kind()) {
+            throw new RuntimeException(
+                    String.format("operation with different type is unsupported: %s(%s, %s)", op.name(), v1.kind(),
+                                  v2.kind()));
+        } else {
+            switch (v1.kind()) {
+                case DOUBLE_VALUE:
+                    return ExprValueFactory.doubleValue(
+                            op.doubleFunc.apply(v1.numberValue().doubleValue(), v2.numberValue().doubleValue()));
+                case INTEGER_VALUE:
+                    return ExprValueFactory
+                            .integerValue(
+                                    op.integerFunc.apply(v1.numberValue().intValue(), v2.numberValue().intValue()));
+                default:
+                    throw new RuntimeException(String.format("unsupported operation: %s(%s, %s)", op.name(), v1.kind(),
+                                                             v2.kind()));
+            }
+        }
+    }
+}
```

Comment:
Does it work if the number of parameters in basic arithmetic expressions is indeterminate? eg. `max(field_1) + max(field_2) + max(field_3) + ... `

Reply:
Yes, it will be compile as add(max(field_1),add(max(field_2),max(field_3)).